Fix LoopStatus errors

llewelld opened the pull request:
1. Return DBus errors for incorrect Set parameters: Fixes a SIGSEGV when incorrect parameters were passed when setting LoopStatus and PlaybackStatus via DBus. Provides a separate DBus interface for org.freedesktop.DBus.Properties rather than using the automatic interface, so that errors can be returned when these invalid values are Set.
2. Send correct LoopStatus values. Send "None", "Track" and "Playlist" instead of "LoopNone", "LoopTrack" and "LoopPlaylist" for the LoopStatus values.
3. Use enums internally for LoopStatus and playbackStatus. Rather than storing the loopStatus and playbackStatus as QStrings, this uses the enums internally to ensure valid values are always maintained.
4. Update canControl at initialisation. Ensures that canControl emits a change signal in case it switches from false (the default) to true during initialisation. Otherwise clients won't notice that it's changed.

llewelld committed:
[mpris] Return DBus errors for incorrect Set parameters. Fixes JB#59853

Fixes a SIGSEGV when incorrect parameters were passed when setting
LoopStatus and PlaybackStatus via DBus.

Provides a separate DBus interface for org.freedesktop.DBus.Properties
rather than using the automatic interface, so that errors can be
returned when these invalid values are Set.

llewelld committed:
[mpris] Send correct LoopStatus values. Fixes JB#59854

Send "None", "Track" and "Playlist" instead of "LoopNone", "LoopTrack"
and "LoopPlaylist" for the LoopStatus values.

inzanity commented on src/mpris.cpp:
```diff
@@ -17,8 +17,75 @@
+QString MprisPrivate::LoopStatusEnumToString(Mpris::LoopStatus value) {
+    switch (value) {
+    case Mpris::LoopNone:
+        return QLatin1String("None");
```
Would QStringLiteral() be better here, as the value is passed in a QString.

llewelld replied:
Thanks for the comments @inzanity. I agree, I've changed it to `QStringLiteral()`.

inzanity commented on src/mprispropertiesadaptor.cpp:
```diff
@@ -0,0 +1,239 @@
+        ReplyPropertyNotFoundError(interface_name, "");
+    }
+
+    for (const QString &property_name : getMap.keys()) {
```
Using iterator would avoid the construction of the QList and value lookups.

llewelld replied:
Yes, good idea. I've changed this too.

inzanity commented on src/mpris.cpp:
```diff
@@ -17,8 +17,75 @@
+QString MprisPrivate::PlaybackStatusEnumToString(Mpris::PlaybackStatus value) {
+    switch (value) {
+    case Mpris::Stopped:
+        return QLatin1String("Stopped");
```
Same here as in LoopStatus...

llewelld replied:
Thanks; changed.

inzanity commented on src/mprispropertiesadaptor.cpp:
```diff
@@ -0,0 +1,241 @@
+    QMap<QString, QString> getMap;
+    QVariant result;
+
+    if (interface_name == QStringLiteral("org.mpris.MediaPlayer2.Player")) {
```
I guess QLatin1String() is usually used in these cases (I guess it mostly just takes less memory)

llewelld replied:
Yes, true. Thanks, I've changed it.

inzanity commented on src/mprispropertiesadaptor.cpp:
```diff
@@ -0,0 +1,241 @@
+    }
+
+    if (getMap.contains(property_name)) {
+        result = Get(getMap.value(property_name));
```
Using `QMap::find()` would remove the need for double lookup.

llewelld replied:
Thanks for pointing this out, I've made some amendments to avoid the double loopkups.

llewelld committed:
[mpris] Use enums internally for LoopStatus and playbackStatus. Fixes JB#59562

Rather than storing the loopStatus and playbackStatus as QStrings, this
uses the enums internally to ensure valid values are always maintained.

Note that this will change the behaviour a little as now setting an
invalid value does not result in change at all. Previously it would
have set the value as is and made the library effectively use a default
value. Internal functions access the values directly through internal
getters and setters.

Co-authored-by: Tomi Leppänen <tomi.leppanen@jolla.com>

llewelld committed:
[mpris] Update canControl at initialisation. Contributes to JB#51154

Ensures that canControl emits a change signal in case it switches from
false (the default) to true during initialisation. Otherwise clients
won't notice that it's changed.

inzanity commented on src/mprispropertiesadaptor.cpp:
```diff
@@ -0,0 +1,240 @@
+{
+    // The property was not found
+    m_playerPrivate->sendErrorReply(QDBusError::UnknownProperty,
+                                    QString::fromLatin1("Property %1%2%3 was not found in object %4")
```
I guess QStringLiteral would be the best choice here.

Although, this is the error path, so the binary size vs performance might lean towards binary size.

llewelld replied:
Yes, I guess you're right: the choice here is between an extra run-time Latin1 -> UTF-16 conversion, and doubling the string size in the binary.

I think this approach here is okay; it's used quite widely elsewhere, e.g. in Qt itself

inzanity commented on src/mprispropertiesadaptor.cpp:
```diff
@@ -0,0 +1,240 @@
+    }
+
+    const QString &getter = getMap.value(property_name);
+    if (!getter.isEmpty()) {
```
I think `.isNull()` would be the most correct check, but this works just as well.

llewelld replied:
Not that it should happen, but this is intended to capture `""` entries in the map as well.

inzanity commented on src/mprispropertiesadaptor_p.h:
```diff
@@ -0,0 +1,89 @@
+    "    <method name=\"GetAll\">\n"
+    "      <arg name=\"interface_name\" type=\"s\" direction=\"in\"/>\n"
+    "      <arg name=\"values\" type=\"a{sv}\" direction=\"out\"/>\n"
+    "      <annotation name=\"org.qtproject.QtDBus.QtTypeName.Out0\" value=\"QVariantMap\"/>\n"
```
Not sure if annotations should be put here, but they shouldn't cause any harm either.

llewelld replied:
You're right, adding the annotations here causes them to appear twice in the introspection. I've removed them from here to avoid that.

Tomin1 commented on src/mpris.cpp:
```diff
@@ -17,8 +17,75 @@
 
 Mpris::Mpris() {}
+
+Mpris::LoopStatus MprisPrivate::LoopStatusStringToEnum(const QString &value, bool *ok) {
```
I wonder about these conversion methods, whether they are better than using QMetaEnum::keyToValue and QMetaEnum::valueToKey. On the other hand it seems that's how it used to work in certain bits of code. I don't know which we should prefer.

llewelld replied:
This is actually the original motivation for this PR: the `LoopStatus` enums were changed, so they no longer match their respective MPRIS-specification strings.

For `LoopStatus` we have to therefore do explicit conversion, but to avoid the same mistake happening again (and also for consistency) I think it makes sense to do the conversion explicitly for `PlaybackStatus` as well.

Tomin1 replied:
You are right, I didn't look carefully enough. :smiley: (Although, I guess `keyToValue()+4` would still work to get `None` from `LoopNone` etc.)

Tomin1 commented on src/mprisplayer_p.h:
```diff
@@ -1,6 +1,6 @@
 /*!
  *
- * Copyright (C) 2021 Jolla Ltd.
+ * Copyright (C) 2023 Jolla Ltd.
```
Shouldn't this be 2021-2023?

llewelld replied:
It should, nicely spotted. I've changed it.

Tomin1 commented on src/mprispropertiesadaptor.cpp:
```diff
@@ -0,0 +1,240 @@
+    }
+
+    QMap<QString, QString>::const_iterator it = getMap.constBegin();
+    while (it != getMap.constEnd()) {
```
I would have probably written this with for loop and auto type:
```
for (auto it = getMap.constBegin(); it != getMap.constEnd(); ++it) {
```
but that's just me. Works like this too. :smiley:

llewelld replied:
In this particular case I prefer to make the typing explicit.

Tomin1 commented on src/mprispropertiesadaptor.cpp:
```diff
@@ -0,0 +1,240 @@
+{
+}
+
+void MprisPropertiesAdaptor::ReplyPropertyNotFoundError(const QString &interface_name, const QString &property_name)
```
If these are methods are not part of the D-Bus interface and there is no existing convention to capitalize the first letter, I would use lower case letter for these methods. Just to avoid confusion with types later in the code.

llewelld replied:
I've lower-cased them.

Tomin1 reviewed (commented):
Some comments. I need to come back to this still though.

pvuorela commented on src/mpris.cpp:
```diff
@@ -17,8 +17,75 @@
 
 Mpris::Mpris() {}
+
+Mpris::LoopStatus MprisPrivate::LoopStatusStringToEnum(const QString &value, bool *ok) {
```
Nit: method starting braces to their own lines. Applies to later ones too.

Also pondering the naming a bit, e.g. stringToLoopStatus() style would drop a few characters with the sort of redundant enum.

llewelld replied:
Brackets and naming changed.

pvuorela commented on src/mprispropertiesadaptor_p.h:
```diff
@@ -0,0 +1,68 @@
+ * This is needed because the standard QDBusAbstractAdaptor
+ * Properties implementation doesn't allow errors to be
+ * returned for the org.freedesktop.DBus.Properties.Set method,
+ * but these are needed buy MPRIS spec.
```
by

llewelld replied:
Thanks! I've changed it.

Tomin1 commented on src/mprispropertiesadaptor.cpp:
```diff
@@ -0,0 +1,240 @@
+}
+
+// Format: {DBus property name, MprisPlayer property name or MprisPlayerPrivate getter() method}
+static const QMap<QString, QString> playerGetMap{
```
Should this have space before `{`?

llewelld replied:
Honestly I'm not sure... I don't think we have conventions to cover this. I took it to be like an initialiser (see the [cpppreference examples](https://en.cppreference.com/w/cpp/language/list_initialization#Example)), which wouldn't normally attract a space.

To get around the uncertainty, I've added a newline, which I think suits our multiline formatting conventions.

Tomin1 commented on src/mprispropertiesadaptor.cpp:
```diff
@@ -0,0 +1,240 @@
+    {"Volume", "volume"},
+};
+
+static const QMap<QString, QString> serviceGetMap{
```
The same here.

Tomin1 commented on src/mprispropertiesadaptor.cpp:
```diff
@@ -0,0 +1,240 @@
+}
+
+// Format: {DBus property name, method to call in MprisPlayerPrivate}
+static const QMap<QString, QString> playerSetMap{
```
And here?

Tomin1 commented on src/mprispropertiesadaptor.cpp:
```diff
@@ -0,0 +1,240 @@
+    {"Volume", "setVolume(double)"},
+};
+
+static const QMap<QString, QString> serviceSetMap{
```
And here?

Tomin1 commented on src/mprispropertiesadaptor.cpp:
```diff
@@ -0,0 +1,240 @@
+        if (method.parameterType(0) == static_cast<int>(value.variant().type())) {
+            success = method.invoke(m_playerPrivate, Qt::DirectConnection,
+                                    QGenericArgument(method.parameterNames()[0],
+                                    value.variant().data()));
```
Should this be indented a little bit more as it's an argument for `QGenericArgument`?

llewelld replied:
Yes it should! Thanks, I've fixed it.

Tomin1 commented on qtdbusextended/dbusextendedabstractinterface.h:
```diff
@@ -90,6 +98,94 @@ private Q_SLOTS:
+
+        int propertyIndex = metaObject()->indexOfProperty(propname);
+
+        if (-1 == propertyIndex) {
```
Interesting choice to have value before the variable in comparison. It's usually the other way around so I wonder why it's like this here.

llewelld replied:
This is a good question; I wouldn't usually do it this way round, but in this case it's mirroring the code in [DBusExtendedAbstractInterface](https://github.com/sailfishos/amber-mpris/blob/92df031f93b5900adb89fa750afbcf33e15afa6f/qtdbusextended/dbusextendedabstractinterface.cpp#L168), which was the basis for these methods.

This usage can be traced back to the [qtdbusextended](https://github.com/sailfishos/qtdbusextended/blob/eaf7e19515e40ddcfcdfe6beff1a708541349046/src/dbusextendedabstractinterface.cpp#L171) project, at which point I guess it was just preference.

My personal opinion is that it's better to mirror the usage, to retain the similarity between the two, but I could go through and change all of them?

Tomin1 replied:
Yeah. I think it's fine, although it doesn't follow any convention that we've had.

Tomin1 commented on qtdbusextended/dbusextendedabstractinterface.h:
```diff
@@ -90,6 +98,94 @@ private Q_SLOTS:
+
+        int propertyIndex = metaObject()->indexOfProperty(propname);
+
+        if (-1 == propertyIndex) {
```
Similarly here.

Tomin1 reviewed (approved):
I had some style related comments to check but otherwise looks good to me and AFAICT it also works. :+1:

Tomin1 commented:
One more thing, one of the commit messages has title:

> [mpris] Send correct LoopStatus values. Fixes JB59854

That is missing `#` character.

llewelld commented:
> That is missing `#` character.

Thanks for your careful review Tomi. I've fixed the commit message too.
